src/libfuncs/enum.rs

```diff
@@ -303,93 +308,36 @@ pub fn build_match<'ctx, 'this>(
             }
 
             // Enum variants.
-            for (i, (block, (payload_ty, payload_layout))) in
+            for (i, (block, (payload_ty, _))) in
                 variant_blocks.into_iter().zip(variant_tys).enumerate()
             {
                 let enum_ty = llvm::r#type::r#struct(context, &[tag_ty, payload_ty], false);
 
-                let payload_val = if type_info.is_memory_allocated(registry) {
-                    let val = block
-                        .append_operation(llvm::load(
-                            context,
-                            entry.argument(0)?.into(),
-                            enum_ty,
-                            location,
-                            LoadStoreOptions::new().align(Some(IntegerAttribute::new(
-                                IntegerType::new(context, 64).into(),
-                                layout.align() as i64,
-                            ))),
-                        ))
-                        .result(0)?
-                        .into();
-
-                    block
-                        .append_operation(llvm::extract_value(
+                let payload_val = match stack_ptr {
+                    Some(stack_ptr) => {
+                        let val = block.load(
                             context,
-                            val,
-                            DenseI64ArrayAttribute::new(context, &[1]),
-                            payload_ty,
                             location,
-                        ))
-                        .result(0)?
-                        .into()
-                } else {
-                    // If the enum is not memory-allocated it means that:
-                    //   - Either it's a C-style enum and all payloads have the same type.
-                    //   - Or the enum only has a single non-memory-allocated variant.
-                    if variant_ids.len() == 1 {
-                        entry.argument(0)?.into()
-                    } else {
-                        assert!(registry.get_type(&variant_ids[i])?.is_zst(registry));
-                        block
-                            .append_operation(llvm::undef(payload_ty, location))
-                            .result(0)?
-                            .into()
+                            stack_ptr,
+                            enum_ty,
+                            Some(layout.align()),
+                        )?;
+                        block.extract_value(context, location, val, payload_ty, 1)?
+                    }
+                    None => {
+                        // If the enum is not memory-allocated it means that:
```

Comment:
I think this kind of comment are really helpful to understand the motivation of the code.

Reply:
That comment was already there.

Reply:
yeah I was just saying that the comment is already great hahah